# **10.1 Введение в словари в Python**

https://stepik.org/lesson/488832/step/1?unit=480068

# **10.2 Основы работы со словарями**

https://stepik.org/lesson/488830/step/1?unit=480066

**Дополните приведенный код так, чтобы он вывел сумму минимального и максимального ключа в словаре my_dict.**


In [ ]:
my_dict = {1.12: 'aa', 67.9: 45, 3.11: 'ccc', 7.9: 'dd', 
           9.2: 'ee', 7.1: 'ff', 0.12: 'qq', 1.91: 'aa', 
           10.12: [1, 2, 3], 99.0: {9, 0, 1}}

print(max(my_dict) + min(my_dict))

99.12


**Дополните приведенный код, чтобы он вывел имена всех пользователей (в алфавитном порядке), чей номер оканчивается на 8.**

Примечание. Имена необходимо вывести на одной строке, разделяя символом пробела.

In [ ]:
users = [{'name': 'Todd', 'phone': '551-1414', 'email': 'todd@gmail.com'},
         {'name': 'Helga', 'phone': '555-1618', 'email': 'helga@mail.net'},
         {'name': 'Olivia', 'phone': '449-3141', 'email': ''},
         {'name': 'LJ', 'phone': '555-2718', 'email': 'lj@gmail.net'},
         {'name': 'Ruslan', 'phone': '422-145-9098', 'email': 'rus-lan.cha@yandex.ru'},
         {'name': 'John', 'phone': '233-421-32', 'email': ''},
         {'name': 'Lara', 'phone': '+7998-676-2532', 'email': 'g.lara89@gmail.com'},
         {'name': 'Alina', 'phone': '+7948-799-2434', 'email': 'ali.ch.b@gmail.com'},
         {'name': 'Robert', 'phone': '420-2011', 'email': ''},
         {'name': 'Riyad', 'phone': '128-8890-128', 'email': 'r.mahrez@mail.net'},
         {'name': 'Khabib', 'phone': '+7995-600-9080', 'email': 'kh.nurmag@gmail.com'},
         {'name': 'Olga', 'phone': '6449-314-1213', 'email': ''},
         {'name': 'Roman', 'phone': '+7459-145-8059', 'email': 'roma988@mail.ru'},
         {'name': 'Maria', 'phone': '12-129-3148', 'email': 'm.sharapova@gmail.com'},
         {'name': 'Fedor', 'phone': '+7445-341-0545', 'email': ''},
         {'name': 'Tim', 'phone': '242-449-3141', 'email': 'timm.ggg@yandex.ru'}]

names = []
for user in users:
    if user['phone'][-1] == '8':
        names.append(user['name'])

print(*sorted(names))

Helga LJ Maria Riyad Ruslan


**Дополните приведенный код, чтобы он вывел имена всех пользователей (в алфавитном порядке), у которых нет информации об электронной почте.** 

Примечание 1. Ключ email может отсутствовать в словаре.

Примечание 2. Имена необходимо вывести на одной строке, разделяя символом пробела.

In [ ]:
users = [{'name': 'Todd', 'phone': '551-1414', 'email': 'todd@gmail.com'},
         {'name': 'Helga', 'phone': '555-1618'},
         {'name': 'Olivia', 'phone': '449-3141', 'email': ''},
         {'name': 'LJ', 'phone': '555-2718', 'email': 'lj@gmail.net'},
         {'name': 'Ruslan', 'phone': '422-145-9098', 'email': 'rus-lan.cha@yandex.ru'},
         {'name': 'John', 'phone': '233-421-32', 'email': ''},
         {'name': 'Lara', 'phone': '+7998-676-2532', 'email': 'g.lara89@gmail.com'},
         {'name': 'Alina', 'phone': '+7948-799-2434'},
         {'name': 'Robert', 'phone': '420-2011', 'email': ''},
         {'name': 'Riyad', 'phone': '128-8890-128', 'email': 'r.mahrez@mail.net'},
         {'name': 'Khabib', 'phone': '+7995-600-9080', 'email': 'kh.nurmag@gmail.com'},
         {'name': 'Olga', 'phone': '6449-314-1213', 'email': ''},
         {'name': 'Roman', 'phone': '+7459-145-8059'},
         {'name': 'Maria', 'phone': '12-129-3148', 'email': 'm.sharapova@gmail.com'},
         {'name': 'Fedor', 'phone': '+7445-341-0545', 'email': ''},
         {'name': 'Tim', 'phone': '242-449-3141', 'email': 'timm.ggg@yandex.ru'}]

names = []
for user in users:
    if 'email' not in user:
        names.append(user['name'])
        continue
    if len(user['email']) == 0:
        names.append(user['name'])

print(*sorted(names))

Alina Fedor Helga John Olga Olivia Robert Roman


**Строковое представление**

Напишите программу, которая будет превращать натуральное число в строку, заменяя все цифры в числе на слова:
- 0 на zero;
- 1 на one;
- 2 на two;
- 3 на three;
- 4 на four;
- 5 на five;
- 6 на six;
- 7 на seven;
- 8 на eight;
- 9 на nine.

Формат входных данных
- На вход программе подается натуральное число.

Формат выходных данных
- Программа должна вывести строковое представление числа.

Примечание. Используйте словарь вместо условного оператора.

In [ ]:
dict_number = {1: 'one', 2: 'two', 3: 'three', 4: 'four', 5: 'five', 6: 'six', 7: 'seven', 8: 'eight',
          9: 'nine', 0: 'zero'}

number = input()
for digit in number:
    print(dict_number[int(digit)], end=' ')

230
two three zero 

**Информация об учебных курсах**

Напишите программу, которая по номеру курса выводит информацию о данном курсе. 

Формат входных данных
- На вход программе подается одна строка – номер курса.

Формат выходных данных
- Программа должна вывести номер курса, затем номер аудитории, имя преподавателя и время проведения курса в соответствии с примерами.

Примечание 1. Используйте словарь вместо условного оператора.

Примечание 2. Для удобного вывода используйте строковый метод format() или f-строки.

In [ ]:
my_dict = {'CS101': (3004, 'Хайнс', '8:00'), 
           'CS102': (4501, 'Альварадо', '9:00'), 
           'CS103': (6755, 'Рич', '10:00'), 
           'NT110': (1244, 'Берк', '11:00'), 
           'CM241': (1411, 'Ли', '13:00')}

course_number = input()
print(f'{course_number}: {my_dict[course_number][0]}, {my_dict[course_number][1]}, {my_dict[course_number][2]}')

CS101
CS101: 3004, Хайнс, 8:00


**Набор сообщений**

На мобильных кнопочных телефонах текстовые сообщения можно отправлять с помощью цифровой клавиатуры. Поскольку с каждой клавишей связано несколько букв, для большинства букв требуется несколько нажатий клавиш. При однократном нажатии цифры генерируется первый символ, указанный для этой клавиши. Нажатие цифры 
2,3,4 или 
5 раз генерирует второй, третий, четвертый или пятый символ клавиши.

Напишите программу, которая отображает нажатия клавиш, необходимые для введенного сообщения.

Формат входных данных
- На вход программе подается одна строка – текстовое сообщение.

Формат выходных данных
- Программа должна вывести нажатия клавиш, необходимых для введенного сообщения.

Примечание 1. Ваша программа должна обрабатывать как прописные, так и строчные буквы.

Примечание 2. Ваша программа должна игнорировать любые символы, не указанные в приведенной выше таблице.

Примечание 3. Nokia 3310, чтобы вспомнить как это было 😄



In [ ]:
keyboard = {
    "1": ".,?!:", "2": "ABC", "3": "DEF",
    "4": "GHI",   "5": "JKL", "6": "MNO",
    "7": "PQRS",  "8": "TUV", "9": "WXYZ",
    "0": " "
}

text = input().upper()
result = ''

for letter in text:
    for key, values in keyboard.items():
        if letter in values:
            result += key * (values.index(letter) + 1)
print(result)

hello
4433555555666


**Код Морзе**

Код Морзе для представления цифр и букв использует тире и точки.

Напишите программу для кодирования текстового сообщения в соответствии с кодом Морзе.

Формат входных данных
- На вход программе подается одна строка – текстовое сообщение.

Формат выходных данных
- Программа должна вывести закодированное с помощью кода Морзе сообщение, оставляя пробел между каждым закодированным символом (последовательностью тире и точек).

Примечание 1. Ваша программа должна игнорировать любые символы, не перечисленные в таблице.

Примечание 2. Код Морзе был разработан в XIX веке и все еще используется, спустя более 
160 лет после создания.

In [ ]:
letters = [c for c in 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789']
morse = ['.-', '-...', '-.-.', '-..', '.', '..-.', '--.', '....', 
         '..', '.---', '-.-', '.-..', '--', '-.', '---', '.--.', 
         '--.-', '.-.', '...', '-', '..-', '...-', '.--', '-..-', 
         '-.--', '--..', '-----', '.----', '..---', '...--', '....-', 
         '.....', '-....', '--...', '---..', '----.']

text = input().upper()
morse_dict = dict(zip(letters, morse))

for letter in text:
    if letter in morse_dict:
        print(morse_dict[letter], end=' ')

SOS
... --- ... 

# **10.3 Методы словарей**

https://stepik.org/lesson/446696/step/1?unit=437002

Дополните приведенный код, чтобы в переменной result хранился словарь, в котором ключи – числа от 
1 до 
15 (включительно), а значения представляют собой квадраты ключей.

**Примечание. Выводить содержимое словаря result не нужно.**

In [ ]:
result = {key: key**2 for key in range(1, 16)}
result

{1: 1,
 2: 4,
 3: 9,
 4: 16,
 5: 25,
 6: 36,
 7: 49,
 8: 64,
 9: 81,
 10: 100,
 11: 121,
 12: 144,
 13: 169,
 14: 196,
 15: 225}

Дополните приведенный код так, чтобы он объединил содержимое двух словарей dict1 и dict2 по ключам, складывая значения по одному и тому же ключу, в случае, если ключ присутствует в обоих словарях. Результирующий словарь необходимо присвоить переменной result.

**Примечание. Выводить содержимое словаря result не нужно.**

In [ ]:
dict1 = {'a': 100, 'z': 333, 'b': 200, 'c': 300, 'd': 45, 'e': 98, 't': 76, 'q': 34, 'f': 90, 'm': 230}
dict2 = {'a': 300, 'b': 200, 'd': 400, 't': 777, 'c': 12, 'p': 123, 'w': 111, 'z': 666}

result = {}
tuple_keys = tuple(dict1.keys()) + tuple(dict2.keys())
for key in tuple_keys:
    result[key] = dict1.get(key, 0) + dict2.get(key, 0)

result

{'a': 400,
 'z': 999,
 'b': 400,
 'c': 312,
 'd': 445,
 'e': 98,
 't': 853,
 'q': 34,
 'f': 90,
 'm': 230,
 'p': 123,
 'w': 111}

Дополните приведенный код так, чтобы в переменной result хранился словарь, в котором для каждого символа строки text будет подсчитано количество его вхождений.

**Примечание. Выводить содержимое словаря result не нужно.**

In [ ]:
text = 'footballcyberpunkextraterritorialityconversationalistblockophthalmoscopicinterdependencemamauserfff'

result = {}
for letter in set(text):
    result[letter] = text.count(letter)

result

{'h': 2,
 'u': 2,
 'l': 6,
 'i': 7,
 's': 4,
 'c': 6,
 'b': 3,
 'v': 1,
 'm': 3,
 'f': 4,
 'e': 10,
 'n': 6,
 'd': 2,
 'y': 2,
 'x': 1,
 'p': 4,
 'r': 8,
 't': 9,
 'a': 8,
 'k': 2,
 'o': 9}

Дополните приведенный код, чтобы он вывел наиболее часто встречающееся слово строки s. Если таких слов несколько, должно быть выведено то, что меньше в лексикографическом порядке.

In [ ]:
s = '''orange strawberry barley gooseberry apple apricot barley currant orange 
       melon pomegranate banana banana orange barley apricot plum grapefruit 
       banana quince strawberry barley grapefruit banana grapes melon strawberry 
       apricot currant currant gooseberry raspberry apricot currant orange lime 
       quince grapefruit barley banana melon pomegranate barley banana orange 
       barley apricot plum banana quince lime grapefruit strawberry gooseberry 
       apple barley apricot currant orange melon pomegranate banana banana orange 
       apricot barley plum banana grapefruit banana quince currant orange melon 
       pomegranate barley plum banana quince barley lime grapefruit pomegranate barley'''

words = {}
lst_keys = []

for word in set(s.split()): 
    words[word] = s.count(word)

maxval = max(words.values())
for key in words:
    if words.get(key) == maxval:
        lst_keys.append(key)

print(sorted(lst_keys)[0])

banana


Вам доступен список pets, содержащий информацию о собаках и их владельцах.  Каждый элемент списка – это кортеж вида (кличка собаки, имя владельца, фамилия владельца, возраст владельца).

Дополните приведенный код так, чтобы в переменной result хранился словарь, в котором для каждого владельца будут перечислены его собаки. Ключом словаря должен быть кортеж (имя, фамилия, возраст владельца), а значением – список кличек собак (сохранив исходный порядок следования).

Примечание 1. Не забывайте: кортежи являются неизменяемыми, поэтому могут быть ключами словаря.

Примечание 2. Обратите внимание, что у некоторых владельцев по несколько собак.

Примечание 3. Выводить содержимое словаря result не нужно.

In [ ]:
pets = [('Hatiko', 'Parker', 'Wilson', 50),
        ('Rusty', 'Josh', 'King', 25),
        ('Fido', 'John', 'Smith', 28),
        ('Butch', 'Jake', 'Smirnoff', 18),
        ('Odi', 'Emma', 'Wright', 18),
        ('Balto', 'Josh', 'King', 25),
        ('Barry', 'Josh', 'King', 25),
        ('Snape', 'Hannah', 'Taylor', 40),
        ('Horry', 'Martha', 'Robinson', 73),
        ('Giro', 'Alex', 'Martinez', 65),
        ('Zooma', 'Simon', 'Nevel', 32),
        ('Lassie', 'Josh', 'King', 25),
        ('Chase', 'Martha', 'Robinson', 73),
        ('Ace', 'Martha', 'Williams', 38),
        ('Rocky', 'Simon', 'Nevel', 32)]

result = {}
for row in pets:
    result.setdefault(row[1:], []).append(row[0])
result

{('Parker', 'Wilson', 50): ['Hatiko'],
 ('Josh', 'King', 25): ['Rusty', 'Balto', 'Barry', 'Lassie'],
 ('John', 'Smith', 28): ['Fido'],
 ('Jake', 'Smirnoff', 18): ['Butch'],
 ('Emma', 'Wright', 18): ['Odi'],
 ('Hannah', 'Taylor', 40): ['Snape'],
 ('Martha', 'Robinson', 73): ['Horry', 'Chase'],
 ('Alex', 'Martinez', 65): ['Giro'],
 ('Simon', 'Nevel', 32): ['Zooma', 'Rocky'],
 ('Martha', 'Williams', 38): ['Ace']}

**Самое редкое слово 🌶️**

На вход программе подается строка текста. Напишите программу, которая выводит слово, которое встречается реже всего, без учета регистра. Если таких слов несколько, выведите то, которое меньше в лексикографическом порядке.

Формат входных данных
- На вход программе подается строка текста.

Формат выходных данных
- Программа должна вывести слово (в нижнем регистре), встречаемое реже всего.

Примечание 1. Программа не должна быть чувствительной к регистру, слова apple и Apple должна воспринимать как одинаковые.

Примечание 2. Слово — последовательность букв. Кроме слов в тексте могут присутствовать пробелы и знаки препинания .,!?:;-, которые нужно игнорировать. Других символов в тексте нет.

In [ ]:
s = input().lower()
s = ''.join([letter for letter in s if letter not in '.,!?:;-,'])
words = {}
lst_words = s.split()
for key in set(lst_words):
    words[key] = lst_words.count(key)

print(min(words.items(), key = lambda item: (item[1], item[0]))[0])

I bought two books: a new book and an old book. The new book was more expensive than the old book.	
a


**Исправление дубликатов 🌶️**

На вход программе подается строка, содержащая строки-идентификаторы. Напишите программу, которая исправляет их так, чтобы в результирующей строке не было дубликатов. Для этого необходимо прибавлять к повторяющимся идентификаторам постфикс _n, где n – количество раз, сколько такой идентификатор уже встречался.

Формат входных данных
- На вход программе подается строка текста, содержащая строки-идентификаторы, разделенные символом пробела.

Формат выходных данных
- Программа должна вывести исправленную строку, не содержащую дубликатов сохранив при этом исходный порядок.

In [ ]:
s = input().split()
dct = {}

for key in s:
    if dct.get(key, 0) == 0:
        elem = key
    else:
        elem = key + '_' + str(dct[key])
    dct[key] = dct.get(key, 0) + 1
    print(elem, end=' ')

i am i r o n m a n
i am i_1 r o n m a n_1 

# **10.4 Задачи на словари**

https://stepik.org/lesson/488831/step/1?unit=480067


**Словарь программиста**

Программисты, как вы уже знаете, постоянно учатся, а в общении между собой используют весьма специфический язык. Чтобы систематизировать ваш пополняющийся профессиональный лексикон, мы придумали эту задачу. Напишите программу создания небольшого словаря сленговых программерских выражений, чтобы она потом по запросу возвращала значения из этого словаря.

Формат входных данных
- В первой строке задано одно целое число 
n — количество слов в словаре. В следующих 
n строках записаны слова и их определения, разделенные двоеточием и символом пробела. В следующей строке записано целое число 
m — количество поисковых слов, чье определение нужно вывести. В следующих 
m строках записаны сами слова, по одному на строке.

Формат выходных данных
- Для каждого слова, независимо от регистра символов, если оно присутствует в словаре, необходимо вывести его определение. Если слова в словаре нет, программа должна вывести "Не найдено", без кавычек.

Примечание 1. Мини-словарь для начинающих разработчиков можно посмотреть тут.

Примечание 2. Гарантируется, что в определяемом слове или фразе отсутствует двоеточие (:), следом за которым идёт пробел.

In [ ]:
mydict = {}
for i in range(int(input())):
    arr = input().split(': ')
    arr[0] = arr[0].lower()
    mydict[arr[0]] = arr[1]


result = [mydict.get(input().lower(), 'Не найдено') for _ in range(int(input()))]

for key in result:
    print(key)

**Анаграммы 1**

Анаграмма – слово (словосочетание), образованное путём перестановки букв, составляющих другое слово (или словосочетание). Например, английские слова evil и live – это анаграммы.

На вход программе подаются два слова. Напишите программу, которая определяет, являются ли они анаграммами.

Формат входных данных
- На вход программе подаются два слова, каждое на отдельной строке.

Формат выходных данных
- Программа должна вывести YES если слова являются анаграммами и NO в противном случае.

In [24]:
word1, word2 = input(), input()

dct1 = {}
dct2 = {}
for letter in word1:
    dct1[letter] = dct1.get(letter, 0) + 1

for letter in word2:
    dct2[letter] = dct2.get(letter, 0) + 1

print('YES') if dct1 == dct2 else print('NO')

thing
night
YES


**Анаграммы 2**

На вход программе подаются два предложения. Напишите программу, которая определяет, являются они анаграммами или нет. Ваша программа должна игнорировать регистр символов, знаки препинания и пробелы.

Формат входных данных
- На вход программе подаются два предложения, каждое на отдельной строке.

Формат выходных данных
- Программа должна вывести YES , если предложения – анаграммы и NO в противном случае.

Примечание. Кроме слов в тексте могут присутствовать пробелы и знаки препинания .,!?:;-. Других символов в тексте нет.

In [31]:
words1 = ''.join([x for x in input().lower() if x not in ' .,!?:;-'])
words2 = ''.join([x for x in input().lower() if x not in ' .,!?:;-'])

dct1, dct2 = {}, {}
for key in words1:
    dct1[key] = dct1.get(key, 0) + 1

for key in words2:
    dct2[key] = dct2.get(key, 0) + 1

print('YES') if dct1 == dct2 else print('NO')

Когда увидимся
тогда и свидимся
NO


**Словарь синонимов**

Вам дан словарь, состоящий из пар слов-синонимов. Повторяющихся слов в словаре нет. Напишите программу, которая для одного данного слова определяет его синоним.

Формат входных данных
- На вход программе подается количество пар синонимов 
n. Далее следует 
n строк, каждая строка содержит два слова-синонима. После этого следует одно слово, синоним которого надо найти.

Формат выходных данных
- Программа должна вывести одно слово, синоним введенного.

Примечание 1. Гарантируется, что синоним введенного слова существует в словаре.

Примечание 2. Все слова в словаре начинаются с заглавной буквы.

In [11]:
dct = {}
for _ in range(int(input())):
    key, value = input().split()
    dct[key] = value

word = input()
if word in dct:
    print(dct[word])
else:
    print([key for key, value in dct.items() if value == word][0])

3
Kind Affable
Intellect Mind
Popular Celebrated
Popular
Celebrated


**Страны и города**

На вход программе подается список стран и городов каждой страны. Затем даны названия городов. Напишите программу, которая для каждого города выводит, в какой стране он находится.

Формат входных данных
- Программа получает на вход количество стран 
n. Далее идет 
n строк, каждая строка начинается с названия страны, затем идут названия городов этой страны. В следующей строке записано число 
m, далее идут 
m запросов — названия каких-то 
m городов, из перечисленных выше.

Формат выходных данных
- Программа должна вывести название страны, в которой находится данный город в соответствии с примером.

In [16]:
cities = {}
for _ in range(int(input())):
    value, key = input().split(maxsplit=1)
    cities.update(cities.fromkeys(key.split(), value))

result = []
for _ in range(int(input())):
    result.append(cities.get(input()))

for city in result:
    print(city)

2
Нидерланды Амстердам Гаага Роттердам Алкмар
Германия Берлин Мюнхен Гамбург Дортмунд
4
Амстердам
Алкмар
Гамбург
Гаага
Нидерланды
Нидерланды
Германия
Нидерланды


**Телефонная книга**

Тимур записал телефоны всех своих друзей, чтобы автоматизировать поиск нужного номера.

У каждого из друзей Тимура может быть один или более телефонных номеров. Напишите программу, которая поможет Тимуру находить все номера определённого друга.

Формат входных данных
- В первой строке задано одно целое число 
n — количество номеров телефонов, информацию о которых Тимур сохранил в телефонной книге. В следующих 
n строках заданы телефоны и имена их владельцев через пробел. В следующей строке записано целое число 
m — количество поисковых запросов от Тимура. В следующих 
m строках записаны сами запросы, по одному на строке. Каждый запрос — это имя друга, чьи телефоны Тимур хочет найти.

Формат выходных данных
- Для каждого запроса от Тимура выведите в отдельной строке все телефоны, принадлежащие человеку с этим именем (независимо от регистра имени). Если в телефонной книге нет телефонов человека с таким именем, выведите в соответствующей строке «абонент не найден» (без кавычек).

Примечание 1. Телефоны одного человека выводите в одну строку через пробел в том порядке, в каком они были заданы во входных данных.

Примечание 2. Количество строк в ответе должно быть равно числу 
m.

Примечание 3. Телефон — это несколько цифр, записанных подряд, а имя может состоять из букв русского или английского алфавита. Записи не повторяются.

In [17]:
contacts = {}
for _ in range(int(input())):
    value, key = input().split()
    contacts.setdefault(key.lower(), []).append(value)

for _ in range(int(input())):
    name_contact = input().lower()
    print(*contacts[name_contact]) if name_contact in contacts else print('абонент не найден')

3
79184219577 Женя
79194249271 Руслан
79281234567 Женя
3
Руслан
79194249271
женя
79184219577 79281234567
Рустам
абонент не найден


**Секретное слово**

Напишите программу для расшифровки секретного слова методом частотного анализа.

Формат входных данных
- В первой строке задано зашифрованное слово. Во второй строке задано одно целое число 
n – количество букв в словаре. В следующих 
n строках записано, сколько раз конкретная буква алфавита встречается в этом слове – <буква>: <частота>.

Формат выходных данных
- Программа должна вывести дешифрованное слово.

Примечание. Гарантируется, что частоты букв не повторяются.

In [21]:
word = input()
dct = {}
for _ in range(int(input())):
    value, key = input().split(': ')
    dct[int(key)] = value

for letter in word:
    print(dct.get(word.count(letter), 0), end='')

*!*!*?
3
а: 3
н: 2
с: 1
ананас

# **10.5 Вложенные словари и генераторы словарей**

https://stepik.org/lesson/446698/step/1?unit=437004

Дополните приведенный код, используя генератор, так чтобы получить словарь result , в котором ключом будет позиция числа в списке numbers (начиная с 
0), а значением – его квадрат.

Примечание. Выводить содержимое словаря result не нужно.

In [2]:
numbers = [34, 10, -4, 6, 10, 23, -90, 100, 21, -35, -95, 1, 36, -38, -19, 1, 6, 87]

result = {key: value**2 for key, value in enumerate(numbers)}
result

{0: 1156,
 1: 100,
 2: 16,
 3: 36,
 4: 100,
 5: 529,
 6: 8100,
 7: 10000,
 8: 441,
 9: 1225,
 10: 9025,
 11: 1,
 12: 1296,
 13: 1444,
 14: 361,
 15: 1,
 16: 36,
 17: 7569}

Дополните приведенный код, используя генератор, чтобы получить словарь result, состоящий из всех элементов словаря colors, кроме тех, у которых значением является None.

Примечание. Выводить содержимое словаря result не нужно.

In [5]:
colors = {'c1': 'Red', 'c2': 'Grey', 'c3': None, 'c4': 'Green', 'c5': 'Yellow', 
          'c6': 'Pink', 'c7': 'Orange', 'c8': None, 'c9': 'White', 'c10': 'Black', 
          'c11': 'Violet', 'c12': 'Gold', 'c13': None, 'c14': 'Amber', 'c15': 'Azure', 
          'c16': 'Beige', 'c17': 'Bronze', 'c18': None, 'c19': 'Lilac', 'c20': 'Pearl', 
          'c21': None, 'c22': 'Sand', 'c23': None}

result = {key: value for key, value in colors.items() if value is not None}
result

{'c1': 'Red',
 'c2': 'Grey',
 'c4': 'Green',
 'c5': 'Yellow',
 'c6': 'Pink',
 'c7': 'Orange',
 'c9': 'White',
 'c10': 'Black',
 'c11': 'Violet',
 'c12': 'Gold',
 'c14': 'Amber',
 'c15': 'Azure',
 'c16': 'Beige',
 'c17': 'Bronze',
 'c19': 'Lilac',
 'c20': 'Pearl',
 'c22': 'Sand'}

Дополните приведенный код, используя генератор, чтобы получить словарь result, состоящий из всех элементов словаря favorite_numbers , значения которых являются двузначными числами.

Примечание. Выводить содержимое словаря result не нужно.

In [6]:
favorite_numbers = {'timur': 17, 'ruslan': 7, 'larisa': 19, 'roman': 123, 
                    'rebecca': 293, 'ronald': 76, 'dorothy': 62, 'harold': 36, 
                    'matt': 314, 'kim': 451, 'rosaly': 18, 'rustam': 89, 
                    'soltan': 111, 'amir': 654, 'dima': 390, 'amiran': 777, 
                    'geor': 999, 'sveta': 75, 'rita': 909, 'kirill': 404, 
                    'olga': 271, 'anna': 55, 'madlen': 876}

result = {key: value for key, value in favorite_numbers.items() if 9 < value < 100}
result

{'timur': 17,
 'larisa': 19,
 'ronald': 76,
 'dorothy': 62,
 'harold': 36,
 'rosaly': 18,
 'rustam': 89,
 'sveta': 75,
 'anna': 55}

Дополните приведенный код, используя генератор, так, чтобы получить словарь result, состоящий из всех элементов словаря months , в которых ключ и значение поменялись местами.

Примечание. Выводить содержимое словаря result не нужно.

In [7]:
months = {1: 'January', 2: 'February', 3: 'March', 4: 'April', 5: 'May', 
          6: 'June', 7: 'July', 8: 'August', 9: 'September', 10: 'October', 
          11: 'November', 12: 'December'}

result = {value: key for key, value in months.items()}
result

{'January': 1,
 'February': 2,
 'March': 3,
 'April': 4,
 'May': 5,
 'June': 6,
 'July': 7,
 'August': 8,
 'September': 9,
 'October': 10,
 'November': 11,
 'December': 12}

В переменной s хранится строка пар число:слово. Дополните приведенный код, используя генератор, чтобы получить словарь result , в котором числа будут ключами, а слова – значениями.

Примечание 1. Ключи словаря должны быть целыми числами (иметь тип int), значения – строками (иметь тип str).

Примечание 2. Выводить содержимое словаря result не нужно.

In [15]:
s = '1:men 2:kind 90:number 0:sun 34:book 56:mountain 87:wood 54:car 3:island 88:power 7:box 17:star 101:ice'

result = {int(x.split(':')[0]): x.split(':')[1] for x in s.split()}
result

{1: 'men',
 2: 'kind',
 90: 'number',
 0: 'sun',
 34: 'book',
 56: 'mountain',
 87: 'wood',
 54: 'car',
 3: 'island',
 88: 'power',
 7: 'box',
 17: 'star',
 101: 'ice'}

Используя генератор, дополните приведенный код, чтобы получить словарь result , где ключом будет элемент списка numbers, а значением – отсортированный по возрастанию список всех его делителей начиная с 
1.

Примечание 1. Если бы список numbers имел вид: 
    numbers = [1, 6, 18], 
то результатом был бы словарь

    result = {1: [1], 6: [1, 2, 3, 6], 18: [1, 2, 3, 6, 9, 18]}
    
Примечание 2. Выводить содержимое словаря result не нужно. 

In [16]:
numbers = [34, 10, 4, 6, 10, 23, 90, 100, 21, 35, 95, 1, 36, 38, 19, 1, 6, 87, 1000, 13456, 360]

result = {key: [x for x in range(1, key + 1) if key % x == 0] for key in numbers}
result

{34: [1, 2, 17, 34],
 10: [1, 2, 5, 10],
 4: [1, 2, 4],
 6: [1, 2, 3, 6],
 23: [1, 23],
 90: [1, 2, 3, 5, 6, 9, 10, 15, 18, 30, 45, 90],
 100: [1, 2, 4, 5, 10, 20, 25, 50, 100],
 21: [1, 3, 7, 21],
 35: [1, 5, 7, 35],
 95: [1, 5, 19, 95],
 1: [1],
 36: [1, 2, 3, 4, 6, 9, 12, 18, 36],
 38: [1, 2, 19, 38],
 19: [1, 19],
 87: [1, 3, 29, 87],
 1000: [1, 2, 4, 5, 8, 10, 20, 25, 40, 50, 100, 125, 200, 250, 500, 1000],
 13456: [1, 2, 4, 8, 16, 29, 58, 116, 232, 464, 841, 1682, 3364, 6728, 13456],
 360: [1,
  2,
  3,
  4,
  5,
  6,
  8,
  9,
  10,
  12,
  15,
  18,
  20,
  24,
  30,
  36,
  40,
  45,
  60,
  72,
  90,
  120,
  180,
  360]}

Дополните приведенный код, используя генератор, так, чтобы получить словарь result , в котором ключом будет строка – элемент списка words, а значением – список соответствующих кодов ASCII символов данной строки.

Примечание 1. Если бы список words имел вид: words = ['yes', 'hello'], то результатом был бы словарь

    result = {'yes': [121, 101, 115], 'hello': [104, 101, 108, 108, 111]}

Примечание 2. Для получения ASCII кода символа используйте функцию ord().

Примечание 3. Выводить содержимое словаря result не нужно.

In [17]:
words = ['hello', 'bye', 'yes', 'no', 'python', 'apple', 'maybe', 'stepik', 'beegeek']

result = {key: [ord(letter) for letter in key] for key in words}
result

{'hello': [104, 101, 108, 108, 111],
 'bye': [98, 121, 101],
 'yes': [121, 101, 115],
 'no': [110, 111],
 'python': [112, 121, 116, 104, 111, 110],
 'apple': [97, 112, 112, 108, 101],
 'maybe': [109, 97, 121, 98, 101],
 'stepik': [115, 116, 101, 112, 105, 107],
 'beegeek': [98, 101, 101, 103, 101, 101, 107]}

Дополните приведенный код, используя генератор, чтобы получить словарь result, состоящий из всех элементов словаря letters , за исключением тех, ключи которых есть в списке remove_keys.

Примечание. Выводить содержимое словаря result не нужно.

In [19]:
letters = {0: 'A', 1: 'B', 2: 'C', 3: 'D', 4: 'E', 
           5: 'F', 6: 'G', 7: 'H', 8: 'I', 9: 'J', 
           10: 'K', 11: 'L', 12: 'M', 13: 'N', 14: 'O', 
           15: 'P', 16: 'Q', 17: 'R', 18: 'S', 19: 'T', 
           20: 'U', 21: 'V', 22: 'W', 23: 'X', 24: 'Y', 
           26: 'Z'}

remove_keys = [1, 5, 7, 12, 17, 19, 21, 24]

result = {key: value for key, value in letters.items() if key not in remove_keys}
result

{0: 'A',
 2: 'C',
 3: 'D',
 4: 'E',
 6: 'G',
 8: 'I',
 9: 'J',
 10: 'K',
 11: 'L',
 13: 'N',
 14: 'O',
 15: 'P',
 16: 'Q',
 18: 'S',
 20: 'U',
 22: 'W',
 23: 'X',
 26: 'Z'}

В переменной students хранится словарь, содержащий информацию о росте (в см) и весе (в кг) студентов.

Дополните приведенный код, используя генератор, чтобы получить словарь result, состоящий из всех элементов словаря students , где указан рост больше 
167 см, а вес меньше 
75 кг.

Примечание. Выводить содержимое словаря result не нужно.

In [20]:
students = {'Timur': (170, 75), 'Ruslan': (180, 105), 'Soltan': (192, 68), 
            'Roman': (175, 70), 'Madlen': (160, 50), 'Stefani': (165, 70), 
            'Tom': (190, 90), 'Jerry': (180, 87), 'Anna': (172, 67), 
            'Scott': (168, 78), 'John': (186, 79), 'Alex': (195, 120), 
            'Max': (200, 110), 'Barak': (180, 89), 'Donald': (170, 80), 
            'Rustam': (186, 100), 'Alice': (159, 59), 'Rita': (170, 80), 
            'Mary': (175, 69), 'Jane': (190, 80)}

result = {key: value for key, value in students.items() if value[0] > 167 and value[1] < 75}
result

{'Soltan': (192, 68), 'Roman': (175, 70), 'Anna': (172, 67), 'Mary': (175, 69)}

Список tuples содержит кортежи, состоящие из трех чисел.

Дополните приведенный код, используя генератор, чтобы получить словарь result, в котором ключом является первый элемент каждого кортежа, а значением – кортеж из оставшихся двух элементов.

Примечание. Выводить содержимое словаря result не нужно.

In [21]:
tuples = [(1, 2, 3), (4, 5, 6), (7, 8, 9), 
          (10, 11, 12), (13, 14, 15), (16, 17, 18), 
          (19, 20, 21), (22, 23, 24), (25, 26, 27), 
          (28, 29, 30), (31, 32, 33), (34, 35, 36)]

result = {value[0]: value[1:] for value in tuples}
result

{1: (2, 3),
 4: (5, 6),
 7: (8, 9),
 10: (11, 12),
 13: (14, 15),
 16: (17, 18),
 19: (20, 21),
 22: (23, 24),
 25: (26, 27),
 28: (29, 30),
 31: (32, 33),
 34: (35, 36)}

Даны три списка student_ids, student_names, student_grades, содержащие информацию о студентах.

Дополните приведенный код, используя генератор, так чтобы получить список result, содержащий вложенные словари в соответствии с образцом: 

    [{'S001': {'Camila Rodriguez': 86}}, {'S002': {'Juan Cruz': 98}},...].

Примечание 1. Для параллельной итерации по всем трем спискам одновременно можно использовать встроенную функцию zip().

Примечание 2. Выводить содержимое списка result не нужно.

In [23]:
student_ids = ['S001', 'S002', 'S003', 'S004', 'S005', 'S006', 
               'S007', 'S008', 'S009', 'S010', 'S011', 'S012', 'S013'] 
student_names = ['Camila Rodriguez', 'Juan Cruz', 'Dan Richards', 'Sam Boyle', 
                 'Batista Cesare', 'Francesco Totti', 'Khalid Hussain', 'Ethan Hawke', 
                 'David Bowman', 'James Milner', 'Michael Owen', 'Gary Oldman', 'Tom Hardy'] 
student_grades = [86, 98, 89, 92, 45, 67, 89, 90, 100, 98, 10, 96, 93]

result = [{student_id: {student_name: student_grade}} 
          for student_id, student_name, student_grade 
          in zip(student_ids, student_names, student_grades)]
result

[{'S001': {'Camila Rodriguez': 86}},
 {'S002': {'Juan Cruz': 98}},
 {'S003': {'Dan Richards': 89}},
 {'S004': {'Sam Boyle': 92}},
 {'S005': {'Batista Cesare': 45}},
 {'S006': {'Francesco Totti': 67}},
 {'S007': {'Khalid Hussain': 89}},
 {'S008': {'Ethan Hawke': 90}},
 {'S009': {'David Bowman': 100}},
 {'S010': {'James Milner': 98}},
 {'S011': {'Michael Owen': 10}},
 {'S012': {'Gary Oldman': 96}},
 {'S013': {'Tom Hardy': 93}}]